server/services/ewelink/lib/device/index.js

```diff
@@ -0,0 +1,76 @@
+const { EVENTS, WEBSOCKET_MESSAGE_TYPES } = require('../../../../utils/constants');
+const { Error401, Error500 } = require('../../../../utils/httpErrors');
+const { EWELINK_EMAIL_KEY, EWELINK_PASSWORD_KEY, EWELINK_REGION_KEY } = require('../utils/constants');
+const { connect } = require('./connect');
+const { discover } = require('./discover');
+const { poll } = require('./poll');
+const { setValue } = require('./setValue');
+const { status } = require('./status');
+
+/**
+ * @description Add ability to control an eWeLink device.
+ * @param {Object} gladys - Gladys instance.
+ * @param {Object} eweLinkApi - EweLink Client.
+ * @param {string} serviceId - UUID of the service in DB.
+ * @example
+ * const EweLinkHandler = new EweLinkHandler(gladys, client, serviceId);
+ */
+const EweLinkHandler = function EweLinkHandler(gladys, eweLinkApi, serviceId) {
+  this.gladys = gladys;
+  this.EweLinkApi = eweLinkApi;
+  this.serviceId = serviceId;
+
+  this.configured = false;
+  this.connected = false;
+  this.accessToken = '';
+  this.apiKey = '';
+};
+
+/**
+ * @description Throw error if EweLinkApi call response has error.
+ * @param {Object} response - EweLinkApi call response.
+ * @param {boolean} emit - True to emit message.
+ * @param {boolean} config - True to reset config.
+ * @example
+ * const EweLinkHandler = new EweLinkHandler(gladys, client, serviceId);
+ */
+async function throwErrorIfNeeded(response, emit = false, config = false) {
+  if (response.error) {
+    if (response.error === 401) {
+      this.connected = false;
+      this.accessToken = '';
+      this.apiKey = '';
+      if (emit) {
+        this.gladys.event.emit(EVENTS.WEBSOCKET.SEND_ALL, {
+          type: WEBSOCKET_MESSAGE_TYPES.EWELINK.ERROR,
+          payload: response.msg,
+        });
+      }
+      if (config) {
+        await Promise.all([
+          this.gladys.variable.setValue(EWELINK_EMAIL_KEY, '', this.serviceId),
+          this.gladys.variable.setValue(EWELINK_PASSWORD_KEY, '', this.serviceId),
+          this.gladys.variable.setValue(EWELINK_REGION_KEY, '', this.serviceId),
+        ]);
+        this.configured = false;
+      }
+      throw new Error401(`EWeLink error: ${response.msg}`);
```

Comment:
Is a 401 error relevant? I'm affraid Gladys thinks user has been disconnected, not only eWeLink account.
To confirm.

Reply:
Yes +1 Please use another error, like 403 Forbidden, or a 400, but not 401 which will trigger a JWT renew and log out the user

Reply:
Ok, 403 fits me.